lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -613,6 +616,44 @@ void FunctionAnalysisResults::importModel() {
   }
 }
 
+void FunctionAnalysisResults::importDynamicSymbols() {
+  llvm::Function &Root = *M.getFunction("root");
+
+  for (BasicBlock &BB : Root) {
+    if (getType(&BB) != BlockType::JumpTargetBlock)
+      continue;
+
+    auto *SymbolNameString = hasDynamicSymbol(&BB);
+    if (not SymbolNameString)
+      continue;
+
+    auto [Entry, _] = getPC(&*BB.begin());
+
+    // Already imported?
+    if (FunctionsBucket.find(Entry) != FunctionsBucket.end())
+      continue;
+
+    llvm::StringRef SymbolName = extractFromConstantStringPtr(SymbolNameString);
+    revng_assert(SymbolName.size() != 0);
+
+    // TODO: change this to assert, dynamic function must have been imported
+    auto It = Binary->ImportedDynamicFunctions.find(SymbolName.str());
+    if (It == Binary->ImportedDynamicFunctions.end())
+      continue;
+
+    if (not It->Prototype.isValid())
+      It->Prototype = Binary->DefaultPrototype;
```

Comment:
Can you make a function in `Binary.h`.
```
getPrototype(const model::Binary &Binary, const model::ImportedFunction &)
```
?
Otherwise this code will end up duplicated in a lot of places. If we have a central place, the day we decide to refactor this, it will be easy.

Reply:
In `include/revng/EarlyFunctionAnalysis/CallEdge.h`, now there's `inline model::TypePath getPrototype(const model::Binary &Binary, MetaAddress CallerFunctionAddress, MetaAddress CallerBlockAddress, const efa::CallEdge &Edge);` (was previously [this one](https://github.com/revng/revng/blob/develop/include/revng/Model/Binary.h#L131), now moved in `efa` namespace). Are we sure we still want to do another specification of this function?

Reply:
@aleclearmind, what shall we do with this? I think the change is really very minor to move `It->Prototype = Binary->DefaultPrototype;` into a function `getPrototype` (which already exists, btw). Also, there's no `model::ImportedFunction`, I guess you meant `model::DynamicFunction`, how would it be used inside the function? I'm just retrieving the prototype of a dynamic function by its name here.

Reply:
I think we need a function to make sure we don't duplicate code.

Reply:
OK, since we need to query `Binary` by Name, and we are also using the `model::DynamicFunction` (for its attributes), if the prototype is valid, I came up with this:
```cpp
inline std::pair<model::TypePath, const model::DynamicFunction &>
getPrototype(const model::Binary &Binary, std::string Name) {
  model::TypePath Result;

  auto It = Binary.ImportedDynamicFunctions.find(Name);
  if (It != Binary.ImportedDynamicFunctions.end()) {
    Result = It->Prototype;
    if (not Result.isValid())
      Result = Binary.DefaultPrototype;
  }

  return { Result, *It };
}
```
Not sure if returning a pair as such is the best approach though.

Reply:
I wouldn't push the look up of the dynamic function by name here.
I'd just do `getPrototype(const model::Binary &Binary, model::DynamicFunction &)`.

Reply:
Then, we `getPrototype` would simply return `Binary.DefaultPrototype`?

Reply:
```c++
model::TypePath
getPrototype(const model::Binary &Binary, model::DynamicFunction &X) {
  if  (X.Prototype.isValid())
    return X.Prototype;
  else
    return Binary.DefaultPrototype;
}
```

Reply:
LG, fixed.